tests/test_peer_funcs.cpp

```diff
@@ -0,0 +1,157 @@
+    char my_endpoint[MAX_SOCKET_STRING];
+    void *router = test_context_socket (ZMQ_ROUTER);
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_bind (router, "inproc://test_mute"));
+    TEST_ASSERT_SUCCESS_ERRNO (
+    zmq_getsockopt (router, ZMQ_LAST_ENDPOINT, my_endpoint, &len));
+
+
+    void *dealer1 = test_context_socket (ZMQ_DEALER);
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (dealer1, ZMQ_ROUTING_ID, "X",
+									1));
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_connect (dealer1, my_endpoint));
+
+
+    void *dealer2 = test_context_socket (ZMQ_DEALER);
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (dealer2, ZMQ_ROUTING_ID, "Y",
+									1));
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_connect (dealer2, my_endpoint));
+
+    //  Send 2 messages on each dealer* -> router
+    send_string_expect_success (dealer1, "HelloX", 0);
+    send_string_expect_success (dealer1, "HelloX", 0);
+
+    send_string_expect_success (dealer2, "HelloY", 0);
+    send_string_expect_success (dealer2, "HelloY", 0);
+
+    //  Reading 1+1 should give two different identities
+    int i = 0;
+    for (i = 0; i < 2; ++i) {
+	rc = zmq_recv (router, buffer, 255, 0);
+	assert (rc == 1);
+
+    bool d1 = strncmp(buffer, "X", strlen("X")) == 0;
+    rc = zmq_recv (router, buffer, 255, 0);
+    assert (rc == 6);
+	if(d1)
+	    assert (strncmp(buffer, "HelloX", strlen("HelloX")) == 0);
+	else
+	    assert (strncmp(buffer, "HelloY", strlen("HelloY")) == 0);
+    }
+
+    // Mute and unseen routing-id
+    rc = zmq_mute_peer (router, "Z", 1, 0x1);
+    assert (rc == -1);
+    assert (errno == EHOSTUNREACH);
+
+    //  Mute X
+    rc = zmq_mute_peer (router, "X", 1, 0x1);
+    assert (rc == 0);
+
+    //  Now we should only receive from Y
+    rc = zmq_recv (router, buffer, 255, 0);
+    assert (rc == 1);
+
+    assert (strncmp(buffer, "Y", 1) == 0);
+    rc = zmq_recv (router, buffer, 255, 0);
+    assert (strncmp(buffer, "HelloY", strlen("HelloY")) == 0);
+
+    //  This should block / return EAGAIN
+    rc = zmq_recv (router, buffer, 255, ZMQ_DONTWAIT);
+    assert (rc == -1);
+    assert (errno == EAGAIN);
+
+    //  Poller should NOT have IN events
+    zmq_pollitem_t poll_items[1] = {{router, 0, ZMQ_POLLIN, 0}};
+    rc = zmq_poll (poll_items, 1, 10);
+    assert (rc == 0);
+
+    //  Unmute X
+    rc = zmq_mute_peer (router, "X", 1, 0x0);
+    assert (rc == 0);
+
+    //  Poller should have IN events
+    rc = zmq_poll (poll_items, 1, 10);
+    assert (rc == 1);
+
+    //  Now we should only receive from X
+    rc = zmq_recv (router, buffer, 255, 0);
+    assert (rc == 1);
+
+    assert (strncmp(buffer, "X", 1) == 0);
+    rc = zmq_recv (router, buffer, 255, 0);
+    assert (strncmp(buffer, "HelloX", strlen("HelloX")) == 0);
+
+    //  Nothing left on the buffer
+    rc = zmq_recv (router, buffer, 255, ZMQ_DONTWAIT);
+    assert (rc == -1);
+    assert (errno == EAGAIN);
+
+    //  Poller should NOT have IN events
+    rc = zmq_poll (poll_items, 1, 10);
+    assert (rc == 0);
+}
```

Comment:
The test sockets should be closed using test_context_socket_close, otherwise a warning will be output by the test

Reply:
Got it